File: lib/workflow_cancel/workflow_cancel_workflow.ex
defmodule WorkflowCancel.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @impl true
  def execute(_context, _input) do
    a = start_activity(EchoActivity, [["Hello World."], 5_000], heartbeat_timeout: 1_000)

    case await_one([{:cancel_request}, a]) do
      {:ok, [%{state: :requested, cause: "cancel_all" = c}, %{state: as}]}
      when as === :cmd or as === :scheduled or as === :started ->
        cancel_activity(a)
        cancel_workflow_execution([[c]])

      {:ok, [%{state: :requested, cause: "cancel_all" = c}, %{}]} ->
        cancel_workflow_execution([[c]])

      {:ok, [%{state: :requested, cause: "cancel_await" = c}, %{}]} ->
        cancel_workflow_execution([[c]])

      {:ok, [%{state: :requested, cause: "cancel_abandon" = c}, %{}]} ->
        await_open_before_close(false)
        cancel_workflow_execution([[c]])

      {:ok, [%{state: :requested, cause: c}, %{}]} ->
        fail_workflow_execution(message: "Unrecognized cancel_workflow reason.", stack_trace: c)

      {:ok, [:noevent, %{}]} ->
        complete_workflow_execution([["Workflow execution cancel not requested."]])
    end
  end
end

File: src/workflow_cancel/workflow_cancel_workflow.erl
-module(workflow_cancel_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

execute(_Context, _Input) ->
    A = start_activity(echo_activity, [[~"Hello World."], 5_000], [{heartbeat_timeout, 1_000}]),
    case await_any([{cancel_request}, A]) of
        {ok, [#{state := requested, cause := ~"cancel_all" = C}, #{state := AS}]} when
            AS =:= cmd; AS =:= scheduled; AS =:= started
        ->
            cancel_activity(A),
            cancel_workflow_execution([[C]]);
        {ok, [#{state := requested, cause := ~"cancel_all" = C}, #{}]} ->
            cancel_workflow_execution([[C]]);
        {ok, [#{state := requested, cause := ~"cancel_await" = C}, #{}]} ->
            cancel_workflow_execution([[C]]);
        {ok, [#{state := requested, cause := ~"cancel_abandon" = C}, #{}]} ->
            await_open_before_close(false),
            cancel_workflow_execution([[C]]);
        {ok, [#{state := requested, cause := C}, #{}]} ->
            fail_workflow_execution([
                {message, "Unrecognized cancel_workflow reason."}, {stack_trace, C}
            ]);
        {ok, [noevent, #{}]} ->
            complete_workflow_execution([[~"Workflow execution cancel not requested."]])
    end.
